# Data Processing

Limpieza y manipulacion de datos para evitar problemas al momento de leerlos en nuestras estructuras.

In [1]:
import pandas as pd

## Payments Dataset

Revisamos como es la data

In [3]:
payments = pd.read_csv("./olist_order_payments_dataset.csv")

payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

Verificamos las columnas

In [4]:
payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')

### Verificar si hay datos que se repiten respecto a la key

In [7]:
payments.shape

(103886, 5)

Cuantos de los ids son unicos? Dado que tenemos 103886 filas

In [8]:
len(payments["order_id"].unique())

99440

Significa que hay datos que se repiten. Alrededor de 5mil.

In [10]:
payments["order_id"] = payments["order_id"].astype(object)


In [11]:
payments.loc[payments.order_id.duplicated(), :]

order_id  payment_sequential payment_type  \
1456    683bf306149bb869980b68d48a1bd6ab                   1  credit_card   
2324    e6a66a8350bb88497954d37688ab123e                   2      voucher   
2393    8e5148bee82a7e42c5f9ba76161dc51a                   1  credit_card   
2414    816ccd9d21435796e8ffa9802b2a782f                   1  credit_card   
2497    2cbcb371aee438c59b722a21d83597e0                   2      voucher   
...                                  ...                 ...          ...   
103778  fd86c80924b4be8fb7f58c4ecc680dae                   1  credit_card   
103817  6d4616de4341417e17978fe57aec1c46                   1  credit_card   
103860  31bc09fdbd701a7a4f9b55b5955b8687                   6      voucher   
103869  c9b01bef18eb84888f0fd071b8413b38                   1  credit_card   
103874  d41f1036712e18ee525939d08b7a1c98                   1  credit_card   

        payment_installments  payment_value  
1456                       1           8.58  
2324                       1          10.51  
2393                       1           0.67  
2414                       1           5.65  
2497                       1           7.80  
...                      ...            ...  
103778                     1          76.10  
103817                     1          19.18  
103860                     1          77.99  
103869                     6         238.16  
103874                     1           3.57  

[4446 rows x 5 columns]

Revisemos uno de los ids que se repiten

In [14]:
payments[payments["order_id"] == "31bc09fdbd701a7a4f9b55b5955b8687"]

order_id  payment_sequential payment_type  \
27995   31bc09fdbd701a7a4f9b55b5955b8687                   3      voucher   
39416   31bc09fdbd701a7a4f9b55b5955b8687                   7      voucher   
43160   31bc09fdbd701a7a4f9b55b5955b8687                   4      voucher   
45492   31bc09fdbd701a7a4f9b55b5955b8687                   1      voucher   
74317   31bc09fdbd701a7a4f9b55b5955b8687                   5      voucher   
82459   31bc09fdbd701a7a4f9b55b5955b8687                   2      voucher   
103860  31bc09fdbd701a7a4f9b55b5955b8687                   6      voucher   

        payment_installments  payment_value  
27995                      1          37.88  
39416                      1           3.48  
43160                      1          40.33  
45492                      1          26.86  
74317                      1          22.28  
82459                      1          55.71  
103860                     1          77.99

Payment sequential funciona como una llave secundaria. Indica orden de la compra. Por ahora mantendremos ambas llaves.

Sin embargo, en caso querramos eliminarlo, podriamos sumar todos los payments_values en una fila.

### Revisar y eliminar filas con valores nulos

In [22]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Vemos que no hay valores nulos, por lo tanto, terminamos con todo lo relacionado a este dataset.

## Products Dataset

Revisamos la data

In [24]:
products = pd.read_csv("./olist_products_dataset.csv")

products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

Verificamos las columnas

In [25]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')

### Verificar si hay datos que se repiten respecto a la key

In [29]:
products.shape

(32951, 9)

Cuantos de los ids son unicos? Dado que tenemos 32951 filas

In [31]:
len(products["product_id"].unique())

32951

Significa que no hay datos que se repitan.

In [34]:
products["product_id"] = products["product_id"].astype(object)


### Revisar y eliminar filas con valores nulos

In [35]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Vemos que hay valores nulos, por lo tanto, los revisaremos y pasaremos a eliminarlos.

In [36]:
products[products.isna().any(axis=1)]

product_id product_category_name  \
105    a41e356c76fab66334f36de622ecbd3a                   NaN   
128    d8dee61c2034d6d075997acef1870e9b                   NaN   
145    56139431d72cd51f19eb9f7dae4d1617                   NaN   
154    46b48281eb6d663ced748f324108c733                   NaN   
197    5fb61f482620cb672f5e586bb132eae9                   NaN   
...                                 ...                   ...   
32515  b0a0c5dd78e644373b199380612c350a                   NaN   
32589  10dbe0fbaa2c505123c17fdc34a63c56                   NaN   
32616  bd2ada37b58ae94cc838b9c0569fecd8                   NaN   
32772  fa51e914046aab32764c41356b9d4ea4                   NaN   
32852  c4ceee876c82b8328e9c293fa0e1989b                   NaN   

       product_name_lenght  product_description_lenght  product_photos_qty  \
105                    NaN                         NaN                 NaN   
128                    NaN                         NaN                 NaN   
145                    NaN                         NaN                 NaN   
154                    NaN                         NaN                 NaN   
197                    NaN                         NaN                 NaN   
...                    ...                         ...                 ...   
32515                  NaN                         NaN                 NaN   
32589                  NaN                         NaN                 NaN   
32616                  NaN                         NaN                 NaN   
32772                  NaN                         NaN                 NaN   
32852                  NaN                         NaN                 NaN   

       product_weight_g  product_length_cm  product_height_cm  \
105               650.0               17.0               14.0   
128               300.0               16.0                7.0   
145               200.0               20.0               20.0   
154             18500.0               41.0               30.0   
197               300.0               35.0                7.0   
...                 ...                ...                ...   
32515            1800.0               30.0               20.0   
32589             800.0               30.0               10.0   
32616             200.0               21.0                8.0   
32772            1300.0               45.0               16.0   
32852             700.0               28.0                3.0   

       product_width_cm  
105                12.0  
128                20.0  
145                20.0  
154                41.0  
197                12.0  
...                 ...  
32515              70.0  
32589              23.0  
32616              16.0  
32772              45.0  
32852              43.0  

[611 rows x 9 columns]

In [40]:
(611/products.shape[0]) *100

1.8542684592273377

Vemos que hay casos donde la falta de datos es muy grave, por lo tanto, dado que estos representan menos del 2% de los datos, los eliminamos.

In [43]:
products = products.dropna()
products

product_id              product_category_name  \
0      1e9e8ef04dbcff4541ed26657ea517e5                         perfumaria   
1      3aa071139cb16b67ca9e5dea641aaa2f                              artes   
2      96bd76ec8810374ed1b65e291975717f                      esporte_lazer   
3      cef67bcfe19066a932b7673e239eb23d                              bebes   
4      9dc1a7de274444849c219cff195d0b71              utilidades_domesticas   
...                                 ...                                ...   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
0                     40.0                       287.0                 1.0   
1                     44.0                       276.0                 1.0   
2                     46.0                       250.0                 1.0   
3                     27.0                       261.0                 1.0   
4                     37.0                       402.0                 4.0   
...                    ...                         ...                 ...   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
0                 225.0               16.0               10.0   
1                1000.0               30.0               18.0   
2                 154.0               18.0                9.0   
3                 371.0               26.0                4.0   
4                 625.0               20.0               17.0   
...                 ...                ...                ...   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
0                  14.0  
1                  20.0  
2                  15.0  
3                  26.0  
4                  13.0  
...                 ...  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0  

[32340 rows x 9 columns]

## Exportamos los datos finales

In [ ]:
payments.to_csv('olist_order_payments_dataset.csv',index=False, header=["order_id", "payment_sequential", "payment_type", "payment_installments", "payment_value"])

In [59]:
products.to_csv('olist_products_dataset.csv',index=False, header=["product_id", "product_category_name", "product_name_lenght", "product_description_lenght", "product_photos_qty", "product_weight_g", "product_length_cm", "product_height_cm", "product_width_cm"])

## Pruebas

In [44]:
prueba_df = products.loc[:,["product_id", "product_category_name", "product_name_lenght", "product_description_lenght", "product_photos_qty"]]

In [45]:
prueba_df.to_csv('../prueba.csv',index=False, header=["product_id", "product_category_name", "product_name_lenght", "product_description_lenght", "product_photos_qty"])